File: backend/rag/argument_builder.py
"""Build coherent argument bundles from retrieved evidence cards."""

from __future__ import annotations

import re
import time
from dataclasses import asdict, dataclass, field
from typing import Any

from .mechanism import parse_mechanism
from .query_intent import QueryIntent, parse_query_intent
from .relevance import _highlight_text, _terms

@dataclass(frozen=True)
class ArgumentCluster:
    id: str
    section: str
    thesis: str
    cards: list[dict[str, Any]]
    supporting_claims: list[str]
    confidence: float

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


@dataclass(frozen=True)
class ArgumentBundle:
    query: str
    opponent_claim: str | None
    main_claim: str
    warrants: list[str]
    cards: list[dict[str, Any]]
    clusters: list[ArgumentCluster]
    source_status: str
    uncertainty: str | None = None

    def to_dict(self) -> dict[str, Any]:
        data = asdict(self)
        data["clusters"] = [cluster.to_dict() for cluster in self.clusters]
        return data


@dataclass(frozen=True)
class GeneratedClaim:
    text: str
    supporting_card_ids: list[str]
    source_status: str

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


@dataclass(frozen=True)
class SourceIntegrityReport:
    source_status: str
    valid_citations: list[str] = field(default_factory=list)
    invalid_citations: list[str] = field(default_factory=list)
    generated_claims: list[GeneratedClaim] = field(default_factory=list)

    def to_dict(self) -> dict[str, Any]:
        data = asdict(self)
        data["generated_claims"] = [claim.to_dict() for claim in self.generated_claims]
        return data


class ArgumentBuilder:
    def __init__(self):
        self.last_debug: dict[str, Any] = {}

    def build(
        self,
        query_or_intent: QueryIntent | str,
        cards: list[dict[str, Any]],
        limit: int = 5,
    ) -> ArgumentBundle:
        started_total = time.perf_counter()
        stats: dict[str, Any] = {
            "input_cards": len(cards),
            "unique_evidence": len({_evidence_key(card) for card in cards}),
            "term_cache_hits": 0,
            "term_cache_misses": 0,
            "phrase_cache_hits": 0,
            "phrase_cache_misses": 0,
            "pair_comparisons": 0,
        }
        cache = _BuildCache()
        started = time.perf_counter()
        intent = (
            parse_query_intent(query_or_intent)
            if isinstance(query_or_intent, str)
            else query_or_intent
        )
        timings = {"parse_intent": _elapsed_ms(started)}
        started = time.perf_counter()
        clusters = cluster_arguments(cards, intent, cache=cache, stats=stats)
        timings["cluster"] = _elapsed_ms(started)
        stats["clusters"] = len(clusters)
        started = time.perf_counter()
        selected = select_diverse_cards(
            clusters,
            limit=limit,
            cache=cache,
            stats=stats,
        )
        timings["select"] = _elapsed_ms(started)
        stats["selected_cards"] = len(selected)
        main_cluster = clusters[0] if clusters else None
        started = time.perf_counter()
        warrants = _warrants(selected)
        timings["warrants"] = _elapsed_ms(started)
        started = time.perf_counter()
        main_claim = _main_claim(intent, main_cluster, warrants)
        timings["main_claim"] = _elapsed_ms(started)
        source_status = "BACKFILE-SOURCED" if selected else "ANALYSIS ONLY"
        uncertainty = None if selected else "No retrieved cards passed the relevance gate."
        started = time.perf_counter()
        bundle = ArgumentBundle(
            query=intent.raw_query,
            opponent_claim=intent.opponent_claim,
            main_claim=main_claim,
            warrants=warrants,
            cards=selected,
            clusters=clusters,
            source_status=source_status,
            uncertainty=uncertainty,
        )
        timings["construct"] = _elapsed_ms(started)
        timings["total"] = _elapsed_ms(started_total)
        self.last_debug = {
            "timings": timings,
            "stats": stats,
        }
        return bundle


def cluster_arguments(
    cards: list[dict[str, Any]],
    intent: QueryIntent | None = None,
    cache: "_BuildCache | None" = None,
    stats: dict[str, Any] | None = None,
) -> list[ArgumentCluster]:
    cache = cache or _BuildCache()
    stats = stats if stats is not None else {}
    grouped: list[list[dict[str, Any]]] = []
    for card in cards:
        for group in grouped:
            if _cluster_similarity(card, group, cache, stats) >= 0.22:
                group.append(card)
                break
        else:
            grouped.append([card])

    clusters = []
    for index, group in enumerate(grouped, start=1):
        ordered = sorted(group, key=_card_score, reverse=True)
        clusters.append(
            ArgumentCluster(
                id=f"cluster-{index}",
                section=str(ordered[0].get("section") or ""),
                thesis=_cluster_thesis(ordered),
                cards=ordered,
                supporting_claims=_supporting_claims(ordered),
                confidence=_cluster_confidence(ordered, intent, cache, stats),
            )
        )
    clusters.sort(key=lambda cluster: cluster.confidence, reverse=True)
    return clusters


def select_diverse_cards(
    clusters: list[ArgumentCluster],
    limit: int = 5,
    lambda_relevance: float = 0.75,
    cache: "_BuildCache | None" = None,
    stats: dict[str, Any] | None = None,
) -> list[dict[str, Any]]:
    cache = cache or _BuildCache()
    stats = stats if stats is not None else {}
    cluster_score_by_card = {
        str(card.get("card_id")): cluster.confidence
        for cluster in clusters
        for card in cluster.cards
    }
    candidates = [card for cluster in clusters for card in cluster.cards]
    selected: list[dict[str, Any]] = []
    while candidates and len(selected) < limit:
        best = max(
            candidates,
            key=lambda card: _mmr_score(
                card,
                selected,
                lambda_relevance,
                cluster_score_by_card,
                cache,
                stats,
            ),
        )
        selected.append(best)
        candidates = [card for card in candidates if str(card.get("card_id")) != str(best.get("card_id"))]
    return selected


def validate_sources(answer: str, bundle: ArgumentBundle) -> SourceIntegrityReport:
    allowed = _citation_labels(bundle.cards)
    found = _citations_in_answer(answer, allowed)
    invalid = _unknown_citation_like_spans(answer, allowed)
    source_status = bundle.source_status
    if not bundle.cards:
        source_status = "ANALYSIS ONLY"
    claims = [
        GeneratedClaim(
            text=sentence,
            supporting_card_ids=[str(card.get("card_id")) for card in bundle.cards],
            source_status=source_status,
        )
        for sentence in _sentences(answer)
        if sentence
    ]
    return SourceIntegrityReport(
        source_status=source_status,
        valid_citations=sorted(found),
        invalid_citations=sorted(invalid),
        generated_claims=claims,
    )


def _cluster_thesis(cards: list[dict[str, Any]]) -> str:
    for card in cards:
        tag = str(card.get("tag") or "").strip()
        if tag:
            return tag
    return "Retrieved evidence supports a related response."


def _supporting_claims(cards: list[dict[str, Any]]) -> list[str]:
    claims = []
    for card in cards:
        tag = str(card.get("tag") or "").strip()
        if tag and tag not in claims:
            claims.append(tag)
    return claims


def _cluster_confidence(
    cards: list[dict[str, Any]],
    intent: QueryIntent | None = None,
    cache: "_BuildCache | None" = None,
    stats: dict[str, Any] | None = None,
) -> float:
    cache = cache or _BuildCache()
    stats = stats if stats is not None else {}
    if not cards:
        return 0.0
    top3 = sorted((_card_score(card) for card in cards), reverse=True)[:3]
    average_top3 = sum(top3) / len(top3)
    size_bonus = min(0.12, 0.035 * max(len(cards) - 1, 0))
    authors = {str(card.get("author") or card.get("card_name") or "") for card in cards}
    documents = {str(card.get("document") or "") for card in cards}
    diversity_bonus = min(0.08, 0.02 * (len(authors) - 1) + 0.02 * (len(documents) - 1))
    query_bonus = _query_cluster_bonus(cards, intent, cache, stats)
    return round(min(1.0, average_top3 + size_bonus + diversity_bonus + query_bonus), 3)


def _query_cluster_bonus(
    cards: list[dict[str, Any]],
    intent: QueryIntent | None,
    cache: "_BuildCache",
    stats: dict[str, Any],
) -> float:
    if intent is None:
        return 0.0
    query_text = intent.opponent_claim or intent.search_text or intent.raw_query
    query_terms = _card_like_terms(query_text)
    if not query_terms:
        return 0.0
    cluster_terms = set()
    for card in cards:
        cluster_terms.update(_card_terms(card, cache, stats))
    overlap = len(query_terms & cluster_terms) / len(query_terms)
    return min(0.22, overlap * 0.22)


def _card_score(card: dict[str, Any]) -> float:
    return float(
        card.get("reranker_score")
        or card.get("relevance_score")
        or card.get("retrieval_score")
        or card.get("score")
        or 0
    )


def _mmr_score(
    card: dict[str, Any],
    selected: list[dict[str, Any]],
    lambda_relevance: float,
    cluster_score_by_card: dict[str, float],
    cache: "_BuildCache",
    stats: dict[str, Any],
) -> float:
    relevance = (
        _card_score(card) * 0.6
        + cluster_score_by_card.get(str(card.get("card_id")), 0.0) * 0.4
    )
    redundancy = max(
        (_similarity(card, other, cache, stats) for other in selected),
        default=0.0,
    )
    return lambda_relevance * relevance - (1.0 - lambda_relevance) * redundancy


def _cluster_similarity(
    card: dict[str, Any],
    group: list[dict[str, Any]],
    cache: "_BuildCache",
    stats: dict[str, Any],
) -> float:
    return max(
        (_similarity(card, other, cache, stats) for other in group),
        default=0.0,
    )


def _similarity(
    left: dict[str, Any],
    right: dict[str, Any],
    cache: "_BuildCache",
    stats: dict[str, Any],
) -> float:
    stats["pair_comparisons"] = int(stats.get("pair_comparisons") or 0) + 1
    left_terms = _card_terms(left, cache, stats)
    right_terms = _card_terms(right, cache, stats)
    if not left_terms or not right_terms:
        return 0.0
    lexical = len(left_terms & right_terms) / len(left_terms | right_terms)
    same_author = str(left.get("author") or left.get("card_name") or "") == str(
        right.get("author") or right.get("card_name") or ""
    )
    same_document = str(left.get("document") or "") == str(right.get("document") or "")
    same_section = str(left.get("section") or "") == str(right.get("section") or "")
    shared_phrases = _phrase_terms(left, cache, stats) & _phrase_terms(right, cache, stats)
    mechanism_bonus = 0.18 if same_section and shared_phrases else 0.0
    return min(
        1.0,
        lexical
        + (0.15 if same_author else 0)
        + (0.1 if same_document else 0)
        + mechanism_bonus,
    )


@dataclass
class _BuildCache:
    card_terms: dict[str, set[str]] = field(default_factory=dict)
    phrase_terms: dict[str, set[str]] = field(default_factory=dict)


def _card_terms(
    card: dict[str, Any],
    cache: _BuildCache | None = None,
    stats: dict[str, Any] | None = None,
) -> set[str]:
    cache = cache or _BuildCache()
    stats = stats if stats is not None else {}
    key = _evidence_key(card)
    cached = cache.card_terms.get(key)
    if cached is not None:
        stats["term_cache_hits"] = int(stats.get("term_cache_hits") or 0) + 1
        return cached
    stats["term_cache_misses"] = int(stats.get("term_cache_misses") or 0) + 1
    text = " ".join(
        str(part)
        for part in [
            card.get("section"),
            card.get("tag"),
            _highlight_text(card),
            _bundle_body_text(card),
        ]
        if part
    )
    mechanism = parse_mechanism(text)
    terms = _terms(text) | mechanism.object_groups | mechanism.phrase_concepts
    cache.card_terms[key] = terms
    return terms


def _card_like_terms(text: str) -> set[str]:
    mechanism = parse_mechanism(text)
    return (_terms(text) | mechanism.object_groups | mechanism.phrase_concepts) - mechanism.generic_terms


def _phrase_terms(
    card: dict[str, Any],
    cache: _BuildCache,
    stats: dict[str, Any],
) -> set[str]:
    key = _evidence_key(card)
    cached = cache.phrase_terms.get(key)
    if cached is not None:
        stats["phrase_cache_hits"] = int(stats.get("phrase_cache_hits") or 0) + 1
        return cached
    stats["phrase_cache_misses"] = int(stats.get("phrase_cache_misses") or 0) + 1
    phrases = {term for term in _card_terms(card, cache, stats) if "_" in term}
    cache.phrase_terms[key] = phrases
    return phrases


def _bundle_body_text(card: dict[str, Any]) -> str:
    return str(card.get("body_preview") or card.get("body") or "")[:2000]


def _evidence_key(card: dict[str, Any]) -> str:
    metadata = card.get("metadata") or {}
    for key in ("content_hash", "evidence_id"):
        value = card.get(key) or metadata.get(key)
        if value:
            return f"{key}:{value}"
    return "|".join(
        str(card.get(key) or "")
        for key in ("card_id", "card_name", "citation", "tag")
    ).lower()


def _elapsed_ms(started: float) -> float:
    return (time.perf_counter() - started) * 1000


def _warrants(cards: list[dict[str, Any]]) -> list[str]:
    warrants = []
    for card in cards:
        highlights = card.get("highlights") or []
        text = ""
        if highlights:
            text = str(highlights[0].get("text") or "").strip()
        if not text:
            text = str(card.get("tag") or "").strip()
        if text and text not in warrants:
            warrants.append(text)
    return warrants


def _main_claim(
    intent: QueryIntent,
    cluster: ArgumentCluster | None,
    warrants: list[str],
) -> str:
    if cluster:
        return cluster.thesis
    if intent.opponent_claim:
        return f"No backfile evidence passed the gate for: {intent.opponent_claim}"
    if warrants:
        return warrants[0]
    return "No backfile evidence passed the gate."


def _citation_labels(cards: list[dict[str, Any]]) -> set[str]:
    labels = set()
    for card in cards:
        for key in ("card_name", "citation"):
            value = str(card.get(key) or "").strip()
            if value:
                labels.add(value)
    return labels


def _citations_in_answer(answer: str, allowed: set[str]) -> set[str]:
    return {label for label in allowed if label and label in answer}


def _unknown_citation_like_spans(answer: str, allowed: set[str]) -> set[str]:
    allowed_short = {label.split(",")[0].strip() for label in allowed}
    spans = set(re.findall(r"\b[A-Z][A-Za-z'’-]{2,}\s+(?:\d{4}|[’']\d{2}|\d{2})\b", answer))
    return {span for span in spans if span not in allowed and span not in allowed_short}


def _sentences(answer: str) -> list[str]:
    body = answer.partition("Sources:")[0]
    return [part.strip() for part in re.split(r"(?<=[.!?])\s+", body) if part.strip()]

File: app/backend/hybrid/argument_builder.cpp
#include "argument_builder.hpp"

#include "relevance.hpp"

#include <algorithm>
#include <cmath>
#include <map>
#include <set>
#include <sstream>

namespace sekret::hybrid {
namespace {

double card_score(const RerankedCard& card) {
    if (card.assessment.relevance_score != 0.0) {
        return card.assessment.relevance_score;
    }
    if (card.card.reranker_score != 0.0) {
        return card.card.reranker_score;
    }
    return card.card.retrieval_score;
}

std::set<std::string> card_terms(const RerankedCard& card) {
    std::ostringstream text;
    text << card.card.section << ' ' << card.card.tag << ' ' << highlight_text(card.card)
         << ' ' << (card.card.body.empty() ? card.card.body_preview : card.card.body);
    auto mechanism = parse_mechanism(text.str());
    auto values = terms(text.str());
    values.insert(mechanism.object_groups.begin(), mechanism.object_groups.end());
    values.insert(mechanism.phrase_concepts.begin(), mechanism.phrase_concepts.end());
    return values;
}

std::set<std::string> phrase_terms(const RerankedCard& card) {
    std::set<std::string> phrases;
    for (const auto& term : card_terms(card)) {
        if (term.find('_') != std::string::npos) {
            phrases.insert(term);
        }
    }
    return phrases;
}

double similarity(const RerankedCard& left, const RerankedCard& right) {
    const auto left_terms = card_terms(left);
    const auto right_terms = card_terms(right);
    if (left_terms.empty() || right_terms.empty()) {
        return 0.0;
    }
    std::size_t overlap = 0;
    for (const auto& term : left_terms) {
        if (right_terms.count(term) != 0) {
            ++overlap;
        }
    }
    std::set<std::string> union_terms = left_terms;
    union_terms.insert(right_terms.begin(), right_terms.end());
    const double lexical = static_cast<double>(overlap) / static_cast<double>(union_terms.size());
    const bool same_author = left.card.author == right.card.author && left.card.author.has_value();
    const bool same_document = !left.card.document_name.empty() && left.card.document_name == right.card.document_name;
    const bool same_section = !left.card.section.empty() && left.card.section == right.card.section;
    const auto left_phrases = phrase_terms(left);
    const auto right_phrases = phrase_terms(right);
    bool shared_phrase = false;
    for (const auto& phrase : left_phrases) {
        if (right_phrases.count(phrase) != 0) {
            shared_phrase = true;
            break;
        }
    }
    return std::min(
        1.0,
        lexical
            + (same_author ? 0.15 : 0.0)
            + (same_document ? 0.1 : 0.0)
            + (same_section && shared_phrase ? 0.18 : 0.0)
    );
}

double cluster_similarity(const RerankedCard& card, const std::vector<RerankedCard>& group) {
    double best = 0.0;
    for (const auto& other : group) {
        best = std::max(best, similarity(card, other));
    }
    return best;
}

std::string cluster_thesis(const std::vector<RerankedCard>& cards) {
    for (const auto& card : cards) {
        if (!card.card.tag.empty()) {
            return card.card.tag;
        }
    }
    return "Retrieved evidence supports a related response.";
}

std::vector<std::string> supporting_claims(const std::vector<RerankedCard>& cards) {
    std::vector<std::string> claims;
    for (const auto& card : cards) {
        if (card.card.tag.empty()) {
            continue;
        }
        if (std::find(claims.begin(), claims.end(), card.card.tag) == claims.end()) {
            claims.push_back(card.card.tag);
        }
    }
    return claims;
}

std::set<std::string> card_like_terms(const std::string& text) {
    auto mechanism = parse_mechanism(text);
    auto values = terms(text);
    values.insert(mechanism.object_groups.begin(), mechanism.object_groups.end());
    values.insert(mechanism.phrase_concepts.begin(), mechanism.phrase_concepts.end());
    for (const auto& generic : mechanism.generic_terms) {
        values.erase(generic);
    }
    return values;
}

double query_cluster_bonus(const std::vector<RerankedCard>& cards, const QueryIntent* intent) {
    if (intent == nullptr) {
        return 0.0;
    }
    const auto query_terms = card_like_terms(retrieval_text(*intent));
    if (query_terms.empty()) {
        return 0.0;
    }
    std::set<std::string> cluster_terms;
    for (const auto& card : cards) {
        auto values = card_terms(card);
        cluster_terms.insert(values.begin(), values.end());
    }
    std::size_t overlap = 0;
    for (const auto& term : query_terms) {
        if (cluster_terms.count(term) != 0) {
            ++overlap;
        }
    }
    return std::min(0.22, (static_cast<double>(overlap) / static_cast<double>(query_terms.size())) * 0.22);
}

double cluster_confidence(const std::vector<RerankedCard>& cards, const QueryIntent* intent) {
    if (cards.empty()) {
        return 0.0;
    }
    std::vector<double> scores;
    for (const auto& card : cards) {
        scores.push_back(card_score(card));
    }
    std::sort(scores.begin(), scores.end(), std::greater<>());
    scores.resize(std::min<std::size_t>(scores.size(), 3));
    double sum = 0.0;
    for (double score : scores) {
        sum += score;
    }
    const double average = sum / static_cast<double>(scores.size());
    const double size_bonus = std::min(0.12, 0.035 * static_cast<double>(cards.size() > 0 ? cards.size() - 1 : 0));
    std::set<std::string> authors;
    std::set<std::string> documents;
    for (const auto& card : cards) {
        authors.insert(card.card.author.value_or(card.card.card_name));
        documents.insert(card.card.document_name);
    }
    const double diversity_bonus = std::min(
        0.08,
        0.02 * static_cast<double>(authors.size() > 0 ? authors.size() - 1 : 0)
            + 0.02 * static_cast<double>(documents.size() > 0 ? documents.size() - 1 : 0)
    );
    return std::round(std::min(1.0, average + size_bonus + diversity_bonus + query_cluster_bonus(cards, intent)) * 1000.0) / 1000.0;
}

double mmr_score(
    const RerankedCard& card,
    const std::vector<RerankedCard>& selected,
    double lambda_relevance,
    const std::map<std::string, double>& cluster_score_by_card
) {
    const auto id = card.card.card_id.empty() ? card.card.id : card.card.card_id;
    const double relevance = card_score(card) * 0.6 + cluster_score_by_card.at(id) * 0.4;
    double redundancy = 0.0;
    for (const auto& other : selected) {
        redundancy = std::max(redundancy, similarity(card, other));
    }
    return lambda_relevance * relevance - (1.0 - lambda_relevance) * redundancy;
}

std::vector<std::string> warrants(const std::vector<RerankedCard>& cards) {
    std::vector<std::string> result;
    for (const auto& card : cards) {
        std::string text;
        if (!card.card.highlights.empty()) {
            text = card.card.highlights.front().text;
        }
        if (text.empty()) {
            text = card.card.tag;
        }
        if (!text.empty() && std::find(result.begin(), result.end(), text) == result.end()) {
            result.push_back(text);
        }
    }
    return result;
}

std::string main_claim(const QueryIntent& intent, const ArgumentCluster* cluster, const std::vector<std::string>& warrants) {
    if (cluster != nullptr) {
        return cluster->thesis;
    }
    if (intent.opponent_claim.has_value()) {
        return "No backfile evidence passed the gate for: " + *intent.opponent_claim;
    }
    if (!warrants.empty()) {
        return warrants.front();
    }
    return "No backfile evidence passed the gate.";
}

} // namespace

ArgumentBundle ArgumentBuilder::build(
    const QueryIntent& intent,
    const std::vector<RerankedCard>& cards,
    std::size_t limit
) const {
    auto clusters = cluster_arguments(cards, &intent);
    auto selected = select_diverse_cards(clusters, limit);
    auto bundle_warrants = warrants(selected);
    const auto* main_cluster = clusters.empty() ? nullptr : &clusters.front();

    ArgumentBundle bundle;
    bundle.query = intent.raw_query;
    bundle.opponent_claim = intent.opponent_claim;
    bundle.main_claim = main_claim(intent, main_cluster, bundle_warrants);
    bundle.warrants = std::move(bundle_warrants);
    bundle.cards = std::move(selected);
    bundle.clusters = std::move(clusters);
    bundle.source_status = bundle.cards.empty() ? "ANALYSIS ONLY" : "BACKFILE-SOURCED";
    if (bundle.cards.empty()) {
        bundle.uncertainty = "No retrieved cards passed the relevance gate.";
    }
    return bundle;
}

std::vector<ArgumentCluster> cluster_arguments(
    const std::vector<RerankedCard>& cards,
    const QueryIntent* intent
) {
    std::vector<std::vector<RerankedCard>> groups;
    for (const auto& card : cards) {
        bool placed = false;
        for (auto& group : groups) {
            if (cluster_similarity(card, group) >= 0.22) {
                group.push_back(card);
                placed = true;
                break;
            }
        }
        if (!placed) {
            groups.push_back({card});
        }
    }

    std::vector<ArgumentCluster> clusters;
    for (std::size_t index = 0; index < groups.size(); ++index) {
        auto group = groups[index];
        std::sort(group.begin(), group.end(), [](const RerankedCard& left, const RerankedCard& right) {
            return card_score(left) > card_score(right);
        });
        ArgumentCluster cluster;
        cluster.id = "cluster-" + std::to_string(index + 1);
        cluster.section = group.empty() ? std::string() : group.front().card.section;
        cluster.thesis = cluster_thesis(group);
        cluster.supporting_claims = supporting_claims(group);
        cluster.confidence = cluster_confidence(group, intent);
        cluster.cards = std::move(group);
        clusters.push_back(std::move(cluster));
    }
    std::sort(clusters.begin(), clusters.end(), [](const ArgumentCluster& left, const ArgumentCluster& right) {
        return left.confidence > right.confidence;
    });
    return clusters;
}

std::vector<RerankedCard> select_diverse_cards(
    const std::vector<ArgumentCluster>& clusters,
    std::size_t limit,
    double lambda_relevance
) {
    std::map<std::string, double> cluster_score_by_card;
    std::vector<RerankedCard> candidates;
    for (const auto& cluster : clusters) {
        for (const auto& card : cluster.cards) {
            const auto id = card.card.card_id.empty() ? card.card.id : card.card.card_id;
            cluster_score_by_card[id] = cluster.confidence;
            candidates.push_back(card);
        }
    }

    std::vector<RerankedCard> selected;
    while (!candidates.empty() && selected.size() < limit) {
        auto best = std::max_element(candidates.begin(), candidates.end(), [&](const RerankedCard& left, const RerankedCard& right) {
            return mmr_score(left, selected, lambda_relevance, cluster_score_by_card)
                < mmr_score(right, selected, lambda_relevance, cluster_score_by_card);
        });
        selected.push_back(*best);
        const auto selected_id = best->card.card_id.empty() ? best->card.id : best->card.card_id;
        candidates.erase(
            std::remove_if(candidates.begin(), candidates.end(), [&](const RerankedCard& card) {
                const auto id = card.card.card_id.empty() ? card.card.id : card.card.card_id;
                return id == selected_id;
            }),
            candidates.end()
        );
    }
    return selected;
}

} // namespace sekret::hybrid
